pkg/workqueue/workqueue.go

```diff
@@ -0,0 +1,77 @@
+package workqueue
+
+import (
+	"math"
+	"sync"
+	"time"
+
+	"k8s.io/client-go/util/workqueue"
+)
+
+type FailureRequest struct {
+	// The number of times the request has been attempted
+	Attempts int
+	// The time at which the request was last attempted
+	LastAttempt time.Time
+}
+
+type TypedItemExponentialTimedFailureRateLimiter[T comparable] struct {
+	failuresLock sync.Mutex
+	failures     map[T]FailureRequest
+
+	baseDelay time.Duration
+	maxDelay  time.Duration
+}
+
+func NewExponentialTimedFailureRateLimiter[T comparable](baseDelay time.Duration, maxDelay time.Duration) workqueue.TypedRateLimiter[T] {
+	return &TypedItemExponentialTimedFailureRateLimiter[T]{
+		failures:  map[T]FailureRequest{},
+		baseDelay: baseDelay,
+		maxDelay:  maxDelay,
+	}
+}
+
+func (r *TypedItemExponentialTimedFailureRateLimiter[T]) When(item T) time.Duration {
+	r.failuresLock.Lock()
+	defer r.failuresLock.Unlock()
+
+	if r.failures[item].LastAttempt != (time.Time{}) && time.Since(r.failures[item].LastAttempt) > 2*r.maxDelay {
```

Comment:
Before this line, I would add a check like this:

```go
failreq, ok := r.failures[item]
if !ok {
   return 0
}
```

And then replace every occurrence of `r.failures[item]` with `failreq`